Flaky test: TestCleanupExternalView.test

### Problem
The test is flaky. It may be due to thread leakage.
```
[ERROR] test(org.apache.helix.integration.TestCleanupExternalView)  Time elapsed: 60.628 s  <<< FAILURE!
java.lang.AssertionError: external-view for TestDB0 should be removed, but was: ZnRecord=TestDB0, {BUCKET_SIZE=0, IDEAL_STATE_MODE=AUTO, NUM_PARTITIONS=2, REBALANCE_MODE=SEMI_AUTO, REBALANCE_STRATEGY=DEFAULT, REPLICAS=2, STATE_MODEL_DEF_REF=MasterSlave, STATE_MODEL_FACTORY_NAME=DEFAULT}{TestDB0_0={localhost_12918=SLAVE, localhost_12919=MASTER}, TestDB0_1={localhost_12918=MASTER, localhost_12919=SLAVE}}{}, Stat=Stat {_version=5, _creationTime=1614716622978, _modifiedTime=1614716623193, _ephemeralOwner=0} expected:<true> but was:<false>
        at org.apache.helix.integration.TestCleanupExternalView.test(TestCleanupExternalView.java:118)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestCleanupExternalView.test:118 external-view for TestDB0 should be removed, but was: ZnRecord=TestDB0, {BUCKET_SIZE=0, IDEAL_STATE_MODE=AUTO, NUM_PARTITIONS=2, REBALANCE_MODE=SEMI_AUTO, REBALANCE_STRATEGY=DEFAULT, REPLICAS=2, STATE_MODEL_DEF_REF=MasterSlave, STATE_MODEL_FACTORY_NAME=DEFAULT}{TestDB0_0={localhost_12918=SLAVE, localhost_12919=MASTER}, TestDB0_1={localhost_12918=MASTER, localhost_12919=SLAVE}}{}, Stat=Stat {_version=5, _creationTime=1614716622978, _modifiedTime=1614716623193, _ephemeralOwner=0} expected:<true> but was:<false>


BeforeClass: TestCleanupExternalView called.
START TestCleanupExternalView test at Tue Mar 02 12:23:42 PST 2021
START TestCleanupExternalView_test at Tue Mar 02 12:23:42 PST 2021
pending zk-events in queue: []
3452624 [TestNGInvoker-test()] ERROR org.apache.helix.TestHelper  - verifier time out, consider try longer timeout, stack trace[java.lang.Thread.getStackTrace(Thread.java:1556), org.apache.helix.TestHelper.verify(TestHelper.java:813), org.apache.helix.integration.TestCleanupExternalView.test(TestCleanupExternalView.java:113), sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method), sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62), sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43), java.lang.reflect.Method.invoke(Method.java:498), org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:76), org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:46), org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:37), java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511), java.util.concurrent.FutureTask.run(FutureTask.java:266), java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142), java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617), java.lang.Thread.run(Thread.java:745)]
END TestCleanupExternalView test at Tue Mar 02 12:24:43 PST 2021, took: 60627ms.
AfterClass: TestCleanupExternalView called.
************ SYSTEM Physical Memory:67386290176
************ total memory:3954 free memory:3148
---------- Test Class TestCleanupExternalView thread leakage detected! ---------------
```

--- a/helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java
+++ b/helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java
@@ -361,17 +361,7 @@ public class GenericHelixController implements IdealStateChangeListener, LiveIns
   private void forceRebalance(HelixManager manager, ClusterEventType eventType) {
     NotificationContext changeContext = new NotificationContext(manager);
     changeContext.setType(NotificationContext.Type.CALLBACK);
-    String uid = UUID.randomUUID().toString().substring(0, 8);
-    ClusterEvent event = new ClusterEvent(_clusterName, eventType, uid);
-    event.addAttribute(AttributeName.EVENT_SESSION.name(),
-        changeContext.getManager().getSessionIdIfLead());
-    event.addAttribute(AttributeName.helixmanager.name(), changeContext.getManager());
-    event.addAttribute(AttributeName.changeContext.name(), changeContext);
-    event.addAttribute(AttributeName.eventData.name(), new ArrayList<>());
-    event.addAttribute(AttributeName.AsyncFIFOWorkerPool.name(), _asyncFIFOWorkerPool);
-
-    enqueueEvent(_taskEventQueue, event);
-    enqueueEvent(_eventQueue, event.clone(uid));
+    pushToEventQueues(eventType, changeContext, Collections.EMPTY_MAP);
     logger.info(String
         .format("Controller rebalance pipeline triggered with event type: %s for cluster %s",
             eventType, _clusterName));
@@ -1271,11 +1261,15 @@ public class GenericHelixController implements IdealStateChangeListener, LiveIns
       HelixManager manager = changeContext.getManager();
       HelixDataAccessor accessor = manager.getHelixDataAccessor();
       Builder keyBuilder = accessor.keyBuilder();
+
       PauseSignal pauseSignal = accessor.getProperty(keyBuilder.pause());
       MaintenanceSignal maintenanceSignal = accessor.getProperty(keyBuilder.maintenance());
-      _paused = updateControllerState(changeContext, pauseSignal, _paused);
-      _inMaintenanceMode =
-          updateControllerState(changeContext, maintenanceSignal, _inMaintenanceMode);
+      boolean prevPaused = _paused;
+      boolean prevInMaintenanceMode = _inMaintenanceMode;
+      _paused = updateControllerState(pauseSignal, _paused);
+      _inMaintenanceMode = updateControllerState(maintenanceSignal, _inMaintenanceMode);
+      triggerResumeEvent(changeContext, prevPaused, prevInMaintenanceMode);
+
       enableClusterStatusMonitor(true);
       _clusterStatusMonitor.setEnabled(!_paused);
       _clusterStatusMonitor.setPaused(_paused);
@@ -1469,33 +1463,39 @@ public class GenericHelixController implements IdealStateChangeListener, LiveIns
     }
   }
 
-  private boolean updateControllerState(NotificationContext changeContext, PauseSignal signal,
-      boolean statusFlag) {
+  private boolean updateControllerState(PauseSignal signal, boolean statusFlag) {
     if (signal != null) {
-      // This logic is used for recording first time entering PAUSE/MAINTENCE mode
       if (!statusFlag) {
         statusFlag = true;
+        // This log is recorded for the first time entering PAUSE/MAINTENANCE mode
         logger.info(String.format("controller is now %s",
             (signal instanceof MaintenanceSignal) ? "in maintenance mode" : "paused"));
       }
     } else {
-      if (statusFlag) {
-        statusFlag = false;
-        logger.info("controller is now resumed from paused state");
-        String uid = UUID.randomUUID().toString().substring(0, 8);
-        ClusterEvent event = new ClusterEvent(_clusterName, ClusterEventType.Resume,
-            String.format("%s_%s", uid, Pipeline.Type.DEFAULT.name()));
-        event.addAttribute(AttributeName.EVENT_SESSION.name(),
-            changeContext.getManager().getSessionIdIfLead());
-        event.addAttribute(AttributeName.changeContext.name(), changeContext);
-        event.addAttribute(AttributeName.helixmanager.name(), changeContext.getManager());
-        event.addAttribute(AttributeName.AsyncFIFOWorkerPool.name(), _asyncFIFOWorkerPool);
-        enqueueEvent(_eventQueue, event);
-        enqueueEvent(_taskEventQueue,
-            event.clone(String.format("%s_%s", uid, Pipeline.Type.TASK.name())));
-      }
+      statusFlag = false;
     }
     return statusFlag;
+  }
+
+  /**
+   * Trigger a Resume Event if the cluster is back to activated.
+   * @param changeContext
+   * @param prevPaused the previous paused status.
+   * @param prevInMaintenanceMode the previous in maintenance mode status.
+   */
+  private void triggerResumeEvent(NotificationContext changeContext, boolean prevPaused,
+      boolean prevInMaintenanceMode) {
+    /**
+     * WARNING: the logic here is tricky.
+     * 1. Only resume if not paused. So if the Maintenance mode is removed but the cluster is still
+     * paused, the resume event should not be sent.
+     * 2. Only send resume event if the status is changed back to active. So we don't send multiple
+     * event unnecessarily.
+     */
+    if (!_paused && (prevPaused || (prevInMaintenanceMode && !_inMaintenanceMode))) {
+      pushToEventQueues(ClusterEventType.Resume, changeContext, Collections.EMPTY_MAP);
+      logger.info("controller is now resumed from paused/maintenance state");
+    }
   }
 
   // TODO: refactor this to use common/ClusterEventProcessor.

